support/util/imagemetadata.go

```diff
@@ -39,32 +42,85 @@ type RegistryClientImageMetadataProvider struct{}
 // stored in the cache.
-func (*RegistryClientImageMetadataProvider) ImageMetadata(ctx context.Context, imageRef string, pullSecret []byte, imageContentSource []hyperv1.ImageContentSource) (*dockerv1client.DockerImageConfig, error) {
+func (r *RegistryClientImageMetadataProvider) ImageMetadata(ctx context.Context, imageRef string, pullSecret []byte) (*dockerv1client.DockerImageConfig, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	var repo distribution.Repository
+	var ref *reference.DockerImageReference
 
-	ref, err := reference.Parse(imageRef)
+	parsedImageRef, err := reference.Parse(imageRef)
 	if err != nil {
 		return nil, fmt.Errorf("failed to parse image reference %q: %w", imageRef, err)
 	}
 
-	// Check if we should override the image for another disconneceted image
-	if len(imageContentSource) > 0 {
-		overridedImageRef, err := GetRegistryOverrides(ctx, ref, imageContentSource)
-		if err != nil {
-			return nil, fmt.Errorf("failed to check registry overrides for image reference %q: %w", imageRef, err)
+	// If the image reference contains a digest, immediately look it up in the cache
+	if parsedImageRef.ID != "" {
+		if imageConfigObject, exists := imageMetadataCache.Get(parsedImageRef.ID); exists {
+			return imageConfigObject.(*dockerv1client.DockerImageConfig), nil
 		}
+	}
 
-		if overridedImageRef != nil {
-			ref = *overridedImageRef
+	// Check if we should override the image for another disconnected image
+	if len(r.OpenShiftImageRegistryOverrides) > 0 {
+		for source, mirrors := range r.OpenShiftImageRegistryOverrides {
+			for _, mirror := range mirrors {
+				ref, err = GetRegistryOverrides(ctx, parsedImageRef, source, mirror)
+				if err != nil {
+					log.Info(fmt.Sprintf("failed to find registry override for image reference %q with source, %s, mirror %s: %s", imageRef, source, mirror, err.Error()))
+					continue
+				}
+
+				// If the image reference contains a digest, immediately look it up in the cache
+				if ref.ID != "" {
+					if imageConfigObject, exists := imageMetadataCache.Get(ref.ID); exists {
+						return imageConfigObject.(*dockerv1client.DockerImageConfig), nil
+					}
+				}
+
+				repo, err = getRepository(ctx, *ref, pullSecret)
+				if err != nil {
+					log.Info(fmt.Sprintf("failed to create repository client for %s with source, %s, mirror %s: %s", ref.DockerClientDefaults().RegistryURL(), source, mirror, err.Error()))
+					continue
+				}
+				break
+			}
+			// We found a successful source/mirror combo so break continuing any further source/mirror combos
+			if repo != nil {
+				break
+			}
 		}
+	} else {
+		ref = &parsedImageRef
+		repo, err = getRepository(ctx, *ref, pullSecret)
 	}
 
-	// If the image reference contains a digest, immediately look it up in the cache
-	if ref.ID != "" {
-		if imageConfigObject, exists := imageMetadataCache.Get(string(ref.ID)); exists {
-			return imageConfigObject.(*dockerv1client.DockerImageConfig), nil
+	if repo != nil {
```

Comment:
now here we can also simplify by returning early and moving the main logic outside the if block
```
	if repo == nil {
		return nil, fmt.Errorf("repo not found this is unexpected ...")
	}

	ref.ID = parsedImageRef.ID
	firstManifest, location, err := manifest.FirstManifest(ctx, *ref, repo)
	if err != nil {
		return nil, fmt.Errorf("failed to obtain root manifest for %s: %w", imageRef, err)
	}

	// If the image ref did not contain a digest, attempt looking it up by digest after we've fetched the digest
	if ref.ID == "" {
		if imageConfigObject, exists := imageMetadataCache.Get(string(location.Manifest)); exists {
			return imageConfigObject.(*dockerv1client.DockerImageConfig), nil
		}
	}

	config, _, err := manifest.ManifestToImageConfig(ctx, firstManifest, repo.Blobs(ctx), location)
	if err != nil {
		return nil, fmt.Errorf("failed to obtain image configuration for %s: %w", imageRef, err)
	}
	imageMetadataCache.Add(string(location.Manifest), config)
```

This function really need a unit test so we can do changes with confidence.

Reply:
updated